src/librustc_typeck/astconv.rs

```diff
@@ -970,60 +970,74 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx> + 'o {
     {
         let tcx = self.tcx();
 
-        if trait_bounds.is_empty() {
-            span_err!(tcx.sess, span, E0224,
-                      "at least one non-builtin trait is required for an object type");
-            return tcx.types.err;
-        }
-
         let mut projection_bounds = Vec::new();
+        let mut potential_assoc_types = Vec::new();
         let dummy_self = self.tcx().types.trait_object_dummy_self;
-        let (principal, potential_assoc_types) = self.instantiate_poly_trait_ref(
-            &trait_bounds[0],
-            dummy_self,
-            &mut projection_bounds,
-        );
-        debug!("principal: {:?}", principal);
-
-        for trait_bound in trait_bounds[1..].iter() {
-            // sanity check for non-principal trait bounds
-            self.instantiate_poly_trait_ref(trait_bound,
-                                            dummy_self,
-                                            &mut vec![]);
+        let bound_trait_refs: Vec<_> = trait_bounds
+            .iter()
+            .rev()
+            .map(|trait_bound| {
+                let (trait_ref, cur_potential_assoc_types) = self.instantiate_poly_trait_ref(
+                    trait_bound,
+                    dummy_self,
+                    &mut projection_bounds
+                );
+                potential_assoc_types.extend(cur_potential_assoc_types.into_iter().flatten());
+                (trait_ref, trait_bound.span)
+            }).collect();
+
+        // Expand trait aliases recursively and check that only one regular (non-auto) trait
+        // is used and no 'maybe' bounds are used.
+        let expanded_traits = traits::expand_trait_aliases(tcx, bound_trait_refs.clone());
+        let (mut auto_traits, regular_traits): (Vec<_>, Vec<_>) =
+            expanded_traits.partition(|i| tcx.trait_is_auto(i.trait_ref().def_id()));
+        if regular_traits.len() > 1 {
+            let extra_trait = &regular_traits[1];
+            let mut err = struct_span_err!(tcx.sess, extra_trait.bottom().1, E0225,
+                "only auto traits can be used as additional traits in a trait object");
+            err.label_with_exp_info(extra_trait, "additional non-auto trait");
+            err.span_label(regular_traits[0].top().1, "first non-auto trait");
+            err.emit();
         }
 
-        let (mut auto_traits, trait_bounds) = split_auto_traits(tcx, &trait_bounds[1..]);
-
-        if !trait_bounds.is_empty() {
-            let b = &trait_bounds[0];
-            let span = b.trait_ref.path.span;
-            struct_span_err!(self.tcx().sess, span, E0225,
-                "only auto traits can be used as additional traits in a trait object")
-                .span_label(span, "non-auto additional trait")
-                .emit();
+        if regular_traits.is_empty() && auto_traits.is_empty() {
+            span_err!(tcx.sess, span, E0224,
+                "at least one non-builtin trait is required for an object type");
```

Comment:
Speaking of `?Sized` (but not `Box<dyn >` with nothing at all)... this makes me curious as to why we don't allow just `dyn ?Sized`. Operationally it seems to me that it shouldn't be any different than `trait Foo {}` + `dyn Foo` or similar for just an auto trait?

Reply:
It's probably reasonable to allow in theory, but maybe that's a more pragmatic reason? CC @nikomatsakis @eddyb

Reply:
I don't have a *problem* allowing `dyn ?Sized`; it's nonsensical (there is no `Sized` default to suppress) but doesn't strike me as harmful.

Reply:
Let's allow it then :)

Reply:
On the other hand we do not allow 1) `trait Foo: ... + ?Sized + ...` -- it seems to me that if we allow `dyn ... + ?Sized + ...` we should also allow 1).

Reply:
+ @eddyb  makes a good point that *"I wouldn't touch this because it may allow things we don't want"*.
+ @petrochenkov also notes that allowing `?Sized` *"would imply that you can hide an unsized type under those bounds, which you cannot (misleading to a reader at least)"*.

I think this sways me over towards banning `dyn ... + ?Sized + ...` for now (sorry for the flip flopping!).

Reply:
Was this on discord? Do you have a link to that discussion?

Reply:
@oli-obk  here: https://discordapp.com/channels/442252698964721669/443151225160990732/573121779153895445

Reply:
Right, so in the end @nikomatsakis [agreed that it's confusing](https://discordapp.com/channels/442252698964721669/443151225160990732/573264819260620821). I'd say, since there is enough (people and rationale) speaking against allowing it and only code simplicity speaking for it, we should ban it.